File: docu/ix/C++/source/EloixClientCpp/src/IXAuthentication.cpp
#include "stdafx.h"
#include "IXAPI.h"
#include "RsaEncryption.h"

namespace de { namespace elo { namespace ix { namespace client { 

#define IX_ENCR_PWD_PREFIX L"##RSA##"

static LPCWSTR RELOGIN_ERRORS[] = {
    L"[ELOIX:2001]", // session expired
};


IXAuthentication::IXAuthentication(PClientInfo clientInfo, PIXProperties sessOpts, EAuthenticationType authType)
	: loginResult(new LoginResult())
	, sessOpts(sessOpts)
	, encrPasswords(new IXProperties())
	, ticketBestBefore(0)
	, ticketBestBeforeAlive()
	, authType(authType)
{
	loginResult->setClientInfo(clientInfo);
	loginResult->setUser(PUserInfo(new UserInfo()));
}

IXAuthentication::~IXAuthentication(void)
{
}

void IXAuthentication::authenticate(PClient client1, function<void (bool, BException)> asyncResult) {

	PClient_Indexserver client = byps_ptr_cast<BClient_Indexserver>(client1);

	auto listenerResult = [this, client, asyncResult](bool ignored, BException e) {
		notifyListeners(client, e);
		asyncResult(ignored, e);
	};

	auto outerResult = [this, client, listenerResult](bool ignored, BException e) {
		if (e) {
			listenerResult(ignored, e);
		}
		else {
			invokeInternalAuthenticateThenGetServerinfo(client, listenerResult);
		}
	};

	providePublicEncrKey(client, outerResult);
}

void IXAuthentication::providePublicEncrKey(PClient_Indexserver client, function<void (bool, BException)> asyncResult) {

	auto asyncLogin = [this, client, asyncResult](PLoginResult loginResult, BException e) {
		if (e) {
			asyncResult(false, e);
		}
		else {
			wstring ticket = loginResult->getClientInfo()->getTicket();
			this->publicKey = toUtf8(ticket);
      asyncResult(true, e);
		}
	};

	client->getIXServicePortIF()->login(L"PUBLIC_ENCR_KEY", L"", L"", L"", asyncLogin);
}

void IXAuthentication::invokeInternalAuthenticateThenGetServerinfo(PClient_Indexserver client, function<void (bool, BException)> asyncResult) {

	auto outerResult = [this, client, asyncResult](bool ignored, BException e) {
		if (e) {
			asyncResult(ignored, e);
		}
		else {
			invokeGetServerInfoThenSetSessionOptions(client, asyncResult);
		}
	};

	internalAuthenticate(client, outerResult);
}

void IXAuthentication::invokeGetServerInfoThenSetSessionOptions(PClient_Indexserver client, function<void (bool, BException)> asyncResult) {

	auto outerResult = [this, client, asyncResult](PServerInfo result, BException e) {
		if (e) {
			asyncResult(false, e);
		}
		else {
			this->databaseEngine = result->getDatabaseEngine();
			invokeSetSessionOptionsThenProvidePasswords(client, asyncResult);
		}
	};

	 client->getIXServicePortIF()->getServerInfo(outerResult);
}

void IXAuthentication::invokeSetSessionOptionsThenProvidePasswords(PClient_Indexserver client, function<void (bool, BException)> asyncResult) {

	auto outerResult = [this, client, asyncResult](bool ignored, BException e) {
		if (e) {
			asyncResult(ignored, e);
		}
		else {
			IXProperties::MAP::iterator itpwd = encrPasswords->map().begin();
			provideEncrPasswords(client, itpwd, asyncResult);
		}
	};

	PSessionOptions opts = makeSessionOptsObj();
	client->getIXServicePortIF()->setSessionOptions(opts, outerResult);
}

void IXAuthentication::provideEncrPasswords(PClient_Indexserver client, IXProperties::MAP::iterator itpwd, function<void (bool, BException)> asyncResult) {

	if (itpwd != encrPasswords->map().end()) {

		wstring cryptId = (*itpwd).first;
		wstring pwd = (*itpwd).second;
		wstring pwdEncr = encryptPassword(pwd);
		itpwd++;

		auto outerResult = [this, client, itpwd, asyncResult](bool ignored, BException e) {
			if (e) {
				asyncResult(ignored, e);
			}
			else {
				provideEncrPasswords(client, itpwd, asyncResult);
			}
		};

		client->getIXServicePortIF()->provideCryptPassword(cryptId, pwdEncr, asyncResult);
	}
	else {
		asyncResult(false, BException());
	}
}

PSessionOptions IXAuthentication::makeSessionOptsObj() {
	PSessionOptions opts(new SessionOptions());
	vector<PKeyValue> vec;
	for (IXProperties::MAP::iterator it = sessOpts->map().begin(); it != sessOpts->map().end(); it++) {
		PKeyValue kv(new KeyValue((*it).first, (*it).second));
		vec.push_back(kv);
	}
	opts->setOptions(byps_ptr< BArray1<PKeyValue> >(new BArray1<PKeyValue>(vec)));
	return opts;	
}

wstring IXAuthentication::encryptPassword(const wstring& pwd) {
	string pwd8 = toUtf8(pwd);

	RsaEncryption rsa;
	rsa.SetPublicKey(publicKey.c_str());
	string encrPwd8 = rsa.Encrypt(pwd8.c_str());

	wstringstream ss;
	ss << IX_ENCR_PWD_PREFIX << fromUtf8(encrPwd8);
	
	return ss.str();
}

bool IXAuthentication::needAlive() {
	return !isExpired() && ticketBestBeforeAlive != 0 && ticketBestBeforeAlive < time(NULL);
}

bool IXAuthentication::isExpired() {
	return ticketBestBefore < time(NULL);
}

void IXAuthentication::applyCallCounter(PClientInfo ci) {
  wstringstream wss;
  wss << L"T" << GetCurrentThreadId() << L":" << (++callCounter);
  ci->setCallId(wss.str());
}

void IXAuthentication::getSession(PClient client1, BTYPEID typeId, function<void (PSerializable, BException)> asyncResult) {
	if (needAlive()) {
		ticketBestBeforeAlive = 0;

		auto aliveResult = [this, asyncResult](bool ignored, BException e) {
			if (!e) {
				setTicketValid();
			}

      PLoginResult lr = loginResult;
      PClientInfo ci = lr ? lr->getClientInfo() : PClientInfo();
      if (!lr && e) {
        e = BException(BExceptionC::INTERNAL, L"Missing login result, invalid authentication status.");
      }

      applyCallCounter(ci);
			asyncResult(ci, e);
		};

		PClient_Indexserver client = byps_ptr_cast<BClient_Indexserver>(client1);
    client->getIXServicePortIF()->alive(aliveResult);

	}
	else {
		PClientInfo ci = loginResult ? loginResult->getClientInfo() : PClientInfo();
    applyCallCounter(ci);
		asyncResult(ci, BException());
	}
}

void IXAuthentication::setLoginResult(PLoginResult lr) {
	loginResult = lr;
	if (lr) {
		setTicketValid();
	}
}

void IXAuthentication::setTicketValid() {
	if (loginResult) {
		ticketBestBefore = time(NULL) + loginResult->getTicketLifetime();
		ticketBestBeforeAlive = time(NULL) + loginResult->getTicketLifetime() / 2;
	}
}

bool IXAuthentication::isReloginException(PClient client, BException ex, BTYPEID typeId) {

	// Diese Requests führen das Relogin selbst durch, weil mit der neuen
    // Anmeldung
    // auch eine neue Teilnehmer-ID für den Ereignisbus erstellt wurde.
    if (typeId == 1721831543L // BRequest_IXServicePortIF_sendEvents.serialVersionUID 
		|| 
		typeId == 563568172L // BRequest_IXServicePortIF_createEventBusListener.serialVersionUID
		) {
      return false;
    }

	if (client->getTransport()->isReloginException(ex, typeId)) return true;

	wstring msg = ex.toString();
	for (size_t i = 0; i < ARRAYSIZE(RELOGIN_ERRORS); i++) {
		if (msg.find(RELOGIN_ERRORS[i]) != wstring::npos) {
			return true;
		}
	}

	return false;
}

void IXAuthentication::notifyListeners(PClient_Indexserver client, BException e) {
	for (auto lsn = listeners.begin(); lsn != listeners.end(); lsn++) {
		(*lsn)->setAsyncResult(client, e);
	}
}

EAuthenticationType IXAuthentication::getAuthenticationType() {
  return authType;
}

PLoginResult IXAuthentication::getLoginResult() {
  return loginResult;
}

}}}}

File: docu/ix/.NET/source/EloixClientCS/EloixClientCS/IndexServer/IXAuthentication.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using byps;

namespace EloixClient.IndexServer
{
    public abstract class IXAuthentication : BAuthentication
    {
        private LoginResult loginResult;
        private readonly IXProperties sessOpts;
        private readonly IXProperties encrPasswords;
        private RSAParameters publicKey;
        private String databaseEngine;
        private List<IXAuthenticationListener> listeners = new List<IXAuthenticationListener>();
        protected long ticketBestBefore;
        protected long ticketBestBeforeAlive;
        internal readonly EAuthenticationType authType;

        public IXAuthentication(ClientInfo clientInfo, IXProperties sessOpts, EAuthenticationType authType)
        {
            this.loginResult = new LoginResult();
            this.loginResult.clientInfo = clientInfo;
            this.loginResult.user = new UserInfo();
            this.sessOpts = sessOpts;
            this.encrPasswords = new IXProperties();
            this.authType = authType;
        }

        protected abstract void internalAuthenticate(BClient_Indexserver client, BAsyncResult<bool> asyncResult);

        public LoginResult getLoginResult()
        {
            lock (this)
            {
                return loginResult;
            }
        }

        public void setLoginResult(LoginResult lr)
        {
            lock (this)
            {
                loginResult = lr;
                setTicketValid();
            }
        }

        internal IXProperties getEncrPasswords()
        {
            return encrPasswords;
        }

        public virtual void authenticate(BClient client1, BAsyncResult<Boolean> asyncResult)
        {
            BClient_Indexserver client = (BClient_Indexserver)client1;

            BAsyncResult<bool> listenerResult = (ignored, e) => 
            {
                notifyListeners(client, e);
                asyncResult(ignored, e);
            };

            BAsyncResult<bool> outerResult = (ignored, e) =>
            {
                if (e != null)
                {
                    asyncResult(ignored, e);
                }
                else
                {
                    invokeInternalAuthenticateThenSetGetServerInfo(client, listenerResult);
                }
            };

            providePublicEncrKey(client, outerResult);
        }

        private void providePublicEncrKey(BClient_Indexserver client, BAsyncResult<bool> asyncResult)
        {
            BAsyncResult<LoginResult> asyncLogin = (lr, e) =>
                {
                    if (e != null)
                    {
                        asyncResult(false, e);
                    }
                    else
                    {
                        publicKey = publicKeyFromString(lr.clientInfo.ticket);
                        asyncResult(true, null);
                    }
                };

            client.IXServicePortIF.login("PUBLIC_ENCR_KEY", null, null, null, asyncLogin);
        }

        private void invokeInternalAuthenticateThenSetGetServerInfo(BClient_Indexserver client, BAsyncResult<bool> asyncResult)
        {
            BAsyncResult<bool> outerResult = (ignored, e) =>
                {
                    if (e != null)
                    {
                        asyncResult(false, e);
                    }
                    else
                    {
                        invokeGetServerinfoThenSetSessionOptions(client, asyncResult);
                    }
                };

            internalAuthenticate(client, outerResult);
        }

        private void invokeGetServerinfoThenSetSessionOptions(BClient_Indexserver client, BAsyncResult<bool> asyncResult)
        {
            BAsyncResult<ServerInfo> outerResult = (serverInfo, e) =>
            {
                if (e != null)
                {
                    asyncResult(false, e);
                }
                else
                {
                    databaseEngine = serverInfo.databaseEngine;
                    invokeSetSessionOptionsThenProvidePasswords(client, asyncResult);
                }
            };

            client.IXServicePortIF.getServerInfo(outerResult);
        }

        private void invokeSetSessionOptionsThenProvidePasswords(BClient_Indexserver client, BAsyncResult<bool> asyncResult)
        {
            BAsyncResult<Object> outerResult = (result, e) =>
            {
                if (e != null)
                {
                    asyncResult(false, e);
                }
                else
                {
                    IEnumerator<String> itpwd = encrPasswords.Keys.GetEnumerator();
                    provideEncrPasswords(client, itpwd, asyncResult);
                }
            };

            SessionOptions opts = makeSessOptsObj();
            client.IXServicePortIF.setSessionOptions(opts, outerResult);
        }

        private void provideEncrPasswords(BClient_Indexserver client, IEnumerator<String> itpwd, BAsyncResult<bool> asyncResult)
        {
            if (itpwd.MoveNext())
            {
                try
                {
                    String cryptId = itpwd.Current;
                    String pwd = encrPasswords[cryptId];
                    String pwdEncr = encryptPassword(pwd);

                    BAsyncResult<Object> outerResult = (ignored, e) =>
                    {
                        if (e != null)
                        {
                            asyncResult(false, e);
                        }
                        else
                        {
                            provideEncrPasswords(client, itpwd, asyncResult);
                        }
                    };

                    client.IXServicePortIF.provideCryptPassword(cryptId, pwdEncr, outerResult);
                }
                catch (Exception e)
                {
                    asyncResult(false, e);
                }
            }
            else
            {
                asyncResult(false, null);
            }
        }

        internal string encryptPassword(string s)
        {
            RSACryptoServiceProvider RSA = new RSACryptoServiceProvider();

            RSA.ImportParameters(publicKey);

            byte[] plainText = Encoding.UTF8.GetBytes(s);
            byte[] encrText = RSA.Encrypt(plainText, false);

            String ret = "##RSA##" + Convert.ToBase64String(encrText);
            return ret;
        }

        private SessionOptions makeSessOptsObj()
        {
            SessionOptions opts = new SessionOptions();
            KeyValue[] kvs = opts.options = new KeyValue[sessOpts.Count];
            int i = 0;
            foreach (String key in sessOpts.Keys)
            {
                KeyValue kv = kvs[i++] = new KeyValue();
                kv.key = key;
                kv.value = sessOpts[key];
            }
            return opts;
        }

        protected readonly static String[] RELOGIN_ERRORS = new String[] 
        { 
            // session expired
            "[ELOIX:" + IXExceptionC.INVALID_SESSION + "]", 
        };

        public virtual bool isReloginException(BClient client, Exception ex, int typeId)
        {
            if (typeId == BRequest_IXServicePortIF_sendEvents.serialVersionUID ||
                typeId == BRequest_IXServicePortIF_createEventBusListener.serialVersionUID)
            {
                return false;
            }

            if (client.getTransport().isReloginException(ex, BRegistry.TYPEID_VOID))
            {
                return true;
            }

            String msg = ex.ToString();
            for (int i = 0; i < RELOGIN_ERRORS.Length; i++)
            {
                if (msg.IndexOf(RELOGIN_ERRORS[i]) >= 0)
                {
                    return true;
                }
            }

            return false;
        }

        public virtual void getSession(BClient client, int typeId, BAsyncResult<Object> asyncResult)
        {
            if (needAlive())
            {
                ticketBestBeforeAlive = 0L;
                BAsyncResult<Object> aliveResult = (ignored, e) =>
                {
                    if (e == null)
                    {
                        setTicketValid();
                    }
                    asyncResult(loginResult.clientInfo, e);
                };
                ((BClient_Indexserver)client).IXServicePortIF.alive(aliveResult);
            }
            else
            {
                asyncResult(loginResult.clientInfo, null);
            }
        }

        private void notifyListeners(BClient client, Exception e)
        {
            foreach (IXAuthenticationListener lsn in listeners)
            {
                lsn(client, e);
            }
        }

        public void addListener(IXAuthenticationListener lsn)
        {
            listeners.Add(lsn);
        }

        public void removeListener(IXAuthenticationListener lsn)
        {
            listeners.Remove(lsn);
        }

        private static byte[] stripLeadingZeros(byte[] a)
        {
            int firstByte = 0;
            for (; firstByte < a.Length && a[firstByte] == 0; firstByte++) { }

            if (firstByte == 0) return a;

            byte[] result = new byte[a.Length - firstByte];
            for (int i = 0; i < result.Length; i++)
            {
                result[i] = a[firstByte + i];
            }
            return result;
        }

        private static RSAParameters publicKeyFromString(string s)
        {
            int p = s.IndexOf(" ");
            if (p < 0) throw new InvalidOperationException("Invalid format.");
            string mod = s.Substring(0, p);
            string exp = s.Substring(p);
            RSAParameters keyInfo = new RSAParameters();

            byte[] modBytes = Convert.FromBase64String(mod);
            keyInfo.Modulus = stripLeadingZeros(modBytes);
            keyInfo.Exponent = Convert.FromBase64String(exp);
            return keyInfo;
        }

        private void setTicketValid()
        {
            if (loginResult != null)
            {
                // 1 Sekunde = 10.000.000 Ticks
                ticketBestBefore = DateTime.Now.Ticks + ((long)loginResult.ticketLifetime * 1000 * 1000 * 10);
                ticketBestBeforeAlive = DateTime.Now.Ticks + ((long)loginResult.ticketLifetime * 500 * 1000 * 10);
            }
        }

        public bool isExpired()
        {
            return loginResult != null && loginResult.clientInfo != null && 
                !String.IsNullOrEmpty(loginResult.clientInfo.ticket) &&
                ticketBestBefore < DateTime.Now.Ticks;
        }

        public bool needAlive()
        {
            return !isExpired() && ticketBestBeforeAlive != 0 && ticketBestBeforeAlive < DateTime.Now.Ticks;
        }

        internal string getDatabaseEngine()
        {
            return databaseEngine;
        }
    }
}
